File: SubsonicSwift/Utilities/SubsonicAuth.swift
import Foundation
import CryptoKit

struct SubsonicAuth {
    static let apiVersion = "1.16.1"
    static let clientName = "SubsonicSwift"

    static func generateSalt() -> String {
        let characters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        return String((0..<12).compactMap { _ in characters.randomElement() })
    }

    static func generateToken(password: String, salt: String) -> String {
        let data = Data((password + salt).utf8)
        let hash = Insecure.MD5.hash(data: data)
        return hash.map { String(format: "%02x", $0) }.joined()
    }

    static func authParams(username: String, password: String) -> [String: String] {
        let salt = generateSalt()
        let token = generateToken(password: password, salt: salt)

        return [
            "u": username,
            "t": token,
            "s": salt,
            "v": apiVersion,
            "c": clientName,
            "f": "json"
        ]
    }
}

File: SmashSonic-Android/app/src/main/java/com/smashsonic/data/remote/SubsonicAuth.kt
package com.smashsonic.data.remote

import java.security.MessageDigest

object SubsonicAuth {
    const val API_VERSION = "1.16.1"
    const val CLIENT_NAME = "SmashSonic"

    fun generateSalt(length: Int = 12): String {
        val chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        return (1..length).map { chars.random() }.joinToString("")
    }

    fun generateToken(password: String, salt: String): String {
        val md5 = MessageDigest.getInstance("MD5")
        val digest = md5.digest((password + salt).toByteArray())
        return digest.joinToString("") { "%02x".format(it) }
    }

    fun authParams(username: String, password: String): Map<String, String> {
        val salt = generateSalt()
        val token = generateToken(password, salt)
        return mapOf(
            "u" to username,
            "t" to token,
            "s" to salt,
            "v" to API_VERSION,
            "c" to CLIENT_NAME,
            "f" to "json",
        )
    }
}
